galaxy: make retryable HTTP status codes configurable

nejch opened the pull request:
##### SUMMARY

Extends the current list of server HTTP codes to retry client-side during galaxy installs, some of which have been listed in https://github.com/ansible/galaxy/issues/3086 and the issues linked. This doesn't solve the root cause, which AFAIK should be resolved once Galaxy NG is ready. IMO these still make sense to be included by default but open to ideas.

This list is now similar to what we have in python-gitlab, which we've been extending based on feedback from users over the years:
https://github.com/python-gitlab/python-gitlab/blob/main/gitlab/const.py#L117

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

For the full list of Cloudflare  52x status codes, see:  
https://developers.cloudflare.com/support/troubleshooting/cloudflare-errors/troubleshooting-cloudflare-5xx-errors/#error-520-web-server-returns-an-unknown-error

/cc @dlouzan @fgreinacher

bcoca commented:
First, not all those codes seem to be wise to add to 'retryable' list.

Second, since that can vary by context, why not make this an option instead (defaulting to currrent list), so the user can configure the codes as they see fit?

nejch commented on lib/ansible/galaxy/api.py:
```diff
@@ -54,7 +48,12 @@ def should_retry_error(exception):
     # Since 403 could reflect the actual problem (such as an expired token), we should
     # not retry by default.
-    if isinstance(exception, GalaxyError) and exception.http_code in RETRY_HTTP_ERROR_CODES:
+    try:
+        retry_codes = [int(code) for code in C.GALAXY_RETRY_HTTP_ERROR_CODES]
+    except ValueError as e:
+        raise AnsibleError("Invalid value for HTTP retry code: %s. Only integer values are supported." % e)
```
Not sure if this should be handled here at all or if we assume they'll always be validated before. Just added just in case.

webknjaz replied:
Yeah, this seems like too late for the validation. But if you do keep this, make sure to chain the exception cause explicitly.
```suggestion
        raise AnsibleError("Invalid value for HTTP retry code: %s. Only integer values are supported." % e) from e
```

nejch commented:
> First, not all those codes seem to be wise to add to 'retryable' list.
> 
> Second, since that can vary by context, why not make this an option instead (defaulting to currrent list), so the user can configure the codes as they see fit?

Thanks @bcoca @webknjaz, I added those initially as I saw 520 was on there which also doesn't necessarily mean retryable but I've removed them now :+1: 

I started adding a simple config item, would you be able to take another look if this is going in the right direction? Wasn't 100% sure on the config approach.

webknjaz commented:
/azp run

nejch commented:
> /azp run

@webknjaz CI is green here but I'm not sure if there's anything else to for this PR before the next round of review? :bow:

webknjaz commented on test/units/galaxy/test_api.py:
```diff
@@ -1353,3 +1353,11 @@ def test_clear_cache(cache_dir):
+
+
+def test_should_retry_error_with_invalid_code(monkeypatch):
+    expected = "Invalid value for HTTP retry code"
```
Since it's a regex, it'd be useful to be extra explicit about it:
```suggestion
    expected = r"^Invalid value for HTTP retry code: invalid\. Only integer values are supported\.$"
```

webknjaz commented on lib/ansible/config/base.yml:
```diff
@@ -1510,6 +1510,18 @@ GALAXY_REQUIRED_VALID_SIGNATURE_COUNT:
+  default:
+    - "429" # Too many requests
+    - "520" # Galaxy rate limit error code (Cloudflare unknown error)
+    - "502" # Common error from galaxy that may represent any number of transient backend issues
```
Is it really necessary to use string literals here?
```suggestion
    - 429 # Too many requests
    - 520 # Galaxy rate limit error code (Cloudflare unknown error)
    - 502 # Common error from galaxy that may represent any number of transient backend issues
```

webknjaz commented:
@nejch I'm mostly going through the CI queue to make sure no statuses are stuck and the CI is healthy. Sometimes GitHub thinks it didn't receive some status updates and it needs to be triggered again. Hence the restarts.

You'll have to wait for somebody to do a functional review.

Akasurde commented on lib/ansible/galaxy/api.py:
```diff
@@ -54,7 +48,12 @@ def should_retry_error(exception):
+    try:
+        retry_codes = [int(code) for code in C.GALAXY_RETRY_HTTP_ERROR_CODES]
+    except ValueError as e:
+        raise AnsibleError("Invalid value for HTTP retry code: %s. Only integer values are supported." % e)
```
```suggestion
        raise AnsibleError(f"Invalid value for HTTP retry code: {e}. Valid values are : {', '.join(str(x) for x in retry_codes)}")
```

Akasurde committed:
galaxy: Make retryable HTTP error codes configurable

* Allow users to configure retryable HTTP error codes while
  making Galaxy Server API calls

Signed-off-by: Abhijeet Kasurde <Akasurde@redhat.com>

Akasurde committed:
Fix test

Signed-off-by: Abhijeet Kasurde <Akasurde@redhat.com>

s-hertel commented on lib/ansible/galaxy/api.py:
```diff
@@ -53,7 +47,12 @@ def should_retry_error(exception):
+    try:
+        retryable_http_codes = [int(code) for code in C.GALAXY_RETRY_HTTP_ERROR_CODES]
+    except ValueError as e:
+        raise AnsibleError("Invalid value for HTTP retryable error code. Only integer values are supported.") from e
```
If the configuration is invalid, the user should get a consistent error, regardless of whether should_retry_error happens to execute. `ansible-config validate` should catch the configuration error. @bcoca do you have thoughts on where the type validation should occur?

Maybe we could add option-specific validation toward [the end of ConfigManager.get_config_value_and_origin](https://github.com/ansible/ansible/blob/devel/lib/ansible/config/manager.py#L710)?

```python
        if config == "GALAXY_RETRY_HTTP_ERROR_CODES":
            try:
                for i in range(0, len(value)):
                    http_error_code = int(value[i])
                    if http_error_code < 100 or http_error_code > 999:
                        raise ValueError(f"{http_error_code} must be between 100-999")
                    value[i] = http_error_code
            except ValueError:
                raise AnsibleOptionsError(f'Invalid value {value!r} for config {_get_config_label(plugin_type, plugin_name, config)}.',
                                          help_text='Valid values are 3-digit numbers.')
```

Akasurde replied:
Done.

s-hertel reviewed (commented):
I'm not sure how much of a need there is for this, now that Galaxy is using https://github.com/ansible/galaxy_ng.

This should have a minor_changes [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment).

Akasurde committed:
Review requests

Signed-off-by: Abhijeet Kasurde <Akasurde@redhat.com>

bcoca commented on lib/ansible/config/manager.py:
```diff
@@ -705,6 +705,17 @@ def get_config_value_and_origin(self, config, cfile=None, plugin_type=None, plug
         else:
             raise AnsibleUndefinedConfigEntry(f'No config definition exists for {_get_config_label(plugin_type, plugin_name, config)}.')
 
+        if config == "GALAXY_RETRY_HTTP_ERROR_CODES":
```
config manager should not have setting specific validations, this should be done at the consumer

Akasurde committed:
Review requests II

Signed-off-by: Abhijeet Kasurde <Akasurde@redhat.com>

s-hertel commented on lib/ansible/galaxy/api.py:
```diff
@@ -53,7 +47,24 @@ def should_retry_error(exception):
+            bad_error_codes.append(code)
+            continue
+
+    if bad_error_codes:
+        raise AnsibleError("Invalid value for HTTP retryable error code. "
+                           f"Only integer between 100 to 999 are supported. Found: {', '.join(bad_error_codes)}")
```
This code only runs when an exception occurs, so the validation will look random from the user's perspective.

s-hertel reviewed (changes requested):
I added the TODO in 2.12 ee725846f070fc6b0dd79b5e8c5199ec652faf87, before Galaxy used GalaxyNG and lots of related ansible-galaxy improvements, so I am not sure that there is still a need for this feature.
